src/marqo/tensor_search/tensor_search.py

```diff
@@ -1235,6 +1346,257 @@ def _lexical_search(
+                    "inner_hits": {
+                        "_source": {
+                            "include": ["__chunks.__field_content", "__chunks.__field_name"]
+                        }
+                    },
+                    "query": {
+                        "knn": {
+                            f"{TensorField.chunks}.{vector_field}": {
+                                "vector": vectorised_text,
+                                "k": result_count + offset
+                            }
+                        }
+                    },
+                    "score_mode": "max"
+                }
+            },
+            "_source": {
+                "exclude": ["__chunks.__vector_*"]
+            }
+        }
+
+        field_names = list(index_info.get_text_properties().keys())
+        if attributes_to_retrieve is not None:
+            search_query["_source"] = {"include": attributes_to_retrieve} if len(attributes_to_retrieve) > 0 else False
+
+        if filter_string is not None:
+            search_query["query"]["nested"]["query"]["knn"][f"{TensorField.chunks}.{vector_field}"][
+                "filter"] = {
+                    "query_string": {"query": f"{contextualised_filter}"}
+                }
+        body += [{"index": index_name}, search_query]
+    return body
+
+def bulk_msearch(config: Config, body: List[Dict]):
+    start_search_http_time = timer()
+    try:
+        response = HttpRequests(config).get(path=F"_msearch", body=utils.dicts_to_jsonl(body))
+        end_search_http_time = timer()
+        total_search_http_time = end_search_http_time - start_search_http_time
+        total_os_process_time = response["took"] * 0.001
+        num_responses = len(response["responses"])
+        logger.info(f"search (tensor) roundtrip: took {total_search_http_time:.3f}s to send {num_responses} search queries (roundtrip) to Marqo-os.")
+
+        responses = [r['hits']['hits'] for r in response["responses"]]
+
+    except KeyError as e:
+        # KeyError indicates we have received a non-successful result
+        try:
+            if "index.max_result_window" in response["responses"][0]["error"]["root_cause"][0]["reason"]:
+                raise errors.IllegalRequestedDocCount(
+            "Marqo-OS rejected the response due to too many requested results. "
+            "Try reducing the query's limit parameter") from e
+            elif 'parse_exception' in response["responses"][0]["error"]["root_cause"][0]["reason"]:
+                raise errors.InvalidArgError("Syntax error, could not parse filter string") from e
+            raise errors.BackendCommunicationError(f"Error communicating with Marqo-OS backend:\n{response}")
+        except (KeyError, IndexError) as e2:
+            raise e
+
+    logger.info(f"  search (tensor) Marqo-os processing time: took {total_os_process_time:.3f}s for Marqo-os to execute the search.")
+    return responses
+
+def gather_documents_from_response(resp):
```

Comment:
docstring: this is used to aggregate documents, for cases where the same document is retrieved for multiple field-queries